README.md
![Gear](https://raw.githubusercontent.com/teambition/gear/master/gear.png)

[![CI](https://github.com/teambition/gear/actions/workflows/ci-cover.yml/badge.svg)](https://github.com/teambition/gear/actions/workflows/ci-cover.yml)
[![Codecov](https://codecov.io/gh/teambition/gear/branch/master/graph/badge.svg)](https://codecov.io/gh/teambition/gear)
[![CodeQL](https://github.com/teambition/gear/actions/workflows/codeql.yml/badge.svg)](https://github.com/teambition/gear/actions/workflows/codeql.yml)
[![Go Reference](https://pkg.go.dev/badge/github.com/teambition/gear.svg)](https://pkg.go.dev/github.com/teambition/gear)
[![License](http://img.shields.io/badge/license-mit-blue.svg?style=flat-square)](https://raw.githubusercontent.com/teambition/gear/master/LICENSE)

A lightweight, composable and high performance web service framework for Go.

## Features

- Effective and flexible middlewares flow control, create anything by middleware
- Powerful and smart HTTP error handling
- Trie base gear.Router, as faster as [HttpRouter](https://github.com/julienschmidt/httprouter), support regexp parameters and group routes
- Integrated timeout context.Context
- Integrated response content compress
- Integrated structured logging middleware
- Integrated request body parser
- Integrated signed cookies
- Integrated JSON, JSONP, XML and HTML renderer
- Integrated CORS, Secure, Favicon and Static middlewares
- More useful methods on gear.Context to manipulate HTTP Request/Response
- Run HTTP and gRPC on the same port
- Completely HTTP/2.0 supported

## Documentation

[Go-Documentation](https://godoc.org/github.com/teambition/gear)

## Import

```go
// package gear
import "github.com/teambition/gear"
```

## Design

1. [Server 底层基于原生 net/http 而不是 fasthttp](https://github.com/teambition/gear/blob/master/doc/design.md#1-server-底层基于原生-nethttp-而不是-fasthttp)
1. [通过 gear.Middleware 中间件模式扩展功能模块](https://github.com/teambition/gear/blob/master/doc/design.md#2-通过-gearmiddleware-中间件模式扩展功能模块)
1. [中间件的单向顺序流程控制和级联流程控制](https://github.com/teambition/gear/blob/master/doc/design.md#3-中间件的单向顺序流程控制和级联流程控制)
1. [功能强大，完美集成 context.Context 的 gear.Context](https://github.com/teambition/gear/blob/master/doc/design.md#4-功能强大完美集成-contextcontext-的-gearcontext)
1. [集中、智能、可自定义的错误和异常处理](https://github.com/teambition/gear/blob/master/doc/design.md#5-集中智能可自定义的错误和异常处理)
1. [After Hook 和 End Hook 的后置处理](https://github.com/teambition/gear/blob/master/doc/design.md#6-after-hook-和-end-hook-的后置处理)
1. [Any interface 无限的 gear.Context 状态扩展能力](https://github.com/teambition/gear/blob/master/doc/design.md#7-any-interface-无限的-gearcontext-状态扩展能力)
1. [请求数据的解析和验证](https://github.com/teambition/gear/blob/master/doc/design.md#8-请求数据的解析和验证)

## FAQ

1. [如何从源码自动生成 Swagger v2 的文档？](https://github.com/teambition/gear/blob/master/doc/faq.md#1-如何从源码自动生成-swagger-v2-的文档)
1. [Go 语言完整的应用项目结构最佳实践是怎样的？](https://github.com/teambition/gear/blob/master/doc/faq.md#2-go-语言完整的应用项目结构最佳实践是怎样的)

## Demo

### Hello

https://github.com/teambition/gear/tree/master/example/hello

```go
  app := gear.New()

  // Add logging middleware
  app.UseHandler(logging.Default(true))

  // Add router middleware
  router := gear.NewRouter()

  // try: http://127.0.0.1:3000/hello
  router.Get("/hello", func(ctx *gear.Context) error {
    return ctx.HTML(200, "<h1>Hello, Gear!</h1>")
  })

  // try: http://127.0.0.1:3000/test?query=hello
  router.Otherwise(func(ctx *gear.Context) error {
    return ctx.JSON(200, map[string]any{
      "Host":    ctx.Host,
      "Method":  ctx.Method,
      "Path":    ctx.Path,
      "URI":     ctx.Req.RequestURI,
      "Headers": ctx.Req.Header,
    })
  })
  app.UseHandler(router)
  app.Error(app.Listen(":3000"))
```

### HTTP2 with Push

https://github.com/teambition/gear/tree/master/example/http2

```go
package main

import (
  "net/http"

  "github.com/teambition/gear"
  "github.com/teambition/gear/logging"
  "github.com/teambition/gear/middleware/favicon"
)

// go run example/http2/app.go
// Visit: https://127.0.0.1:3000/
func main() {
  const htmlBody = `
<!DOCTYPE html>
<html>
  <head>
    <link href="/hello.css" rel="stylesheet" type="text/css">
  </head>
  <body>
    <h1>Hello, Gear!</h1>
  </body>
</html>`

  const pushBody = `
h1 {
  color: red;
}
`

  app := gear.New()

  app.UseHandler(logging.Default(true))
  app.Use(favicon.New("./testdata/favicon.ico"))

  router := gear.NewRouter()
  router.Get("/", func(ctx *gear.Context) error {
    ctx.Res.Push("/hello.css", &http.PushOptions{Method: "GET"})
    return ctx.HTML(200, htmlBody)
  })
  router.Get("/hello.css", func(ctx *gear.Context) error {
    ctx.Type("text/css")
    return ctx.End(200, []byte(pushBody))
  })
  app.UseHandler(router)
  app.Error(app.ListenTLS(":3000", "./testdata/out/test.crt", "./testdata/out/test.key"))
}
```

### A CMD tool: static server

https://github.com/teambition/gear/tree/master/example/staticgo

Install it with go:

```sh
go install github.com/teambition/gear/example/staticgo@latest
```

It is a useful CMD tool that serve your local files as web server (support TLS).
You can build `osx`, `linux`, `windows` version with `make build`.

```go
package main

import (
  "flag"

  "github.com/teambition/gear"
  "github.com/teambition/gear/logging"
  "github.com/teambition/gear/middleware/cors"
  "github.com/teambition/gear/middleware/static"
)

var (
  address  = flag.String("addr", "127.0.0.1:3000", `address to listen on.`)
  path     = flag.String("path", "./", `static files path to serve.`)
  certFile = flag.String("certFile", "", `certFile path, used to create TLS static server.`)
  keyFile  = flag.String("keyFile", "", `keyFile path, used to create TLS static server.`)
)

func main() {
  flag.Parse()
  app := gear.New()

  app.UseHandler(logging.Default(true))
  app.Use(cors.New())
  app.Use(static.New(static.Options{Root: *path}))

  logging.Println("staticgo v1.2.0, created by https://github.com/teambition/gear")
  logging.Printf("listen: %s, serve: %s\n", *address, *path)

  if *certFile != "" && *keyFile != "" {
    app.Error(app.ListenTLS(*address, *certFile, *keyFile))
  } else {
    app.Error(app.Listen(*address))
  }
}
```

### A CMD tool: reverse proxy server

https://github.com/teambition/gear/tree/master/example/gearproxy

Install it with go:

```sh
go install github.com/teambition/gear/example/gearproxy@latest
gearproxy -help
```

### HTTP2 & gRPC

https://github.com/teambition/gear/tree/master/example/grpc_server

https://github.com/teambition/gear/tree/master/example/grpc_client

## About Router

[gear.Router](https://godoc.org/github.com/teambition/gear#Router) is a trie base HTTP request handler.
Features:

1. Support named parameter
1. Support regexp
1. Support suffix matching
1. Support multi-router
1. Support router layer middlewares
1. Support fixed path automatic redirection
1. Support trailing slash automatic redirection
1. Automatic handle `405 Method Not Allowed`
1. Automatic handle `OPTIONS` method
1. Best Performance

The registered path, against which the router matches incoming requests, can contain six types of parameters:

| Syntax                 | Description                                     |
| ---------------------- | ----------------------------------------------- |
| `:name`                | named parameter                                 |
| `:name(regexp)`        | named with regexp parameter                     |
| `:name+suffix`         | named parameter with suffix matching            |
| `:name(regexp)+suffix` | named with regexp parameter and suffix matching |
| `:name*`               | named with catch-all parameter                  |
| `::name`               | not named parameter, it is literal `:name`      |

Named parameters are dynamic path segments. They match anything until the next '/' or the path end:

Defined: `/api/:type/:ID`

```md
/api/user/123             matched: type="user", ID="123"
/api/user                 no match
/api/user/123/comments    no match
```

Named with regexp parameters match anything using regexp until the next '/' or the path end:

Defined: `/api/:type/:ID(^\d+$)`

```md
/api/user/123             matched: type="user", ID="123"
/api/user                 no match
/api/user/abc             no match
/api/user/123/comments    no match
```

Named parameters with suffix, such as [Google API Design](https://cloud.google.com/apis/design/custom_methods):

Defined: `/api/:resource/:ID+:undelete`


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file starting from version **v1.0.0**.
This project adheres to [Semantic Versioning](http://semver.org/).

-----
## [1.27.3] - 2023-08-08

**Changed:**

- Add `gear.ThresholdCompress` instead of `github.com/teambition/compressible-go`.
- Update dependencies.

## [1.27.2] - 2023-07-10

**Changed:**

- Add gearproxy as example.

## [1.27.1] - 2023-07-04

**Changed:**

- Update dependencies.
- Add a const `MIMEApplicationCBOR`.


## [1.27.0] - 2023-01-13

**Changed:**

- Improve gear.Context with generics.
- Add `gear.State` for passing information through middleware and handlers.
- Add functions:
  - `func CtxWith[T any](parent context.Context, v *T) context.Context`
  - `func CtxValue[T any](ctx context.Context) *T`
  - `func CtxDoIf[T any, TI isValid[T]](ctx context.Context, fn func(v *T))`

**Fixed:**

- Fix gear.ValuesToStruct for slice type.


## [1.26.0] - 2022-06-30

**Changed:**

- Support h2c, the non-TLS version of HTTP/2

## [1.25.1] - 2022-02-06

**Fixed:**

- Fix favicon middleware.

## [1.25.0] - 2021-09-10

**Changed:**

- update trie-mux mod for issue #54. Your routing may be affected!

## [1.24.2] - 2021-09-10

**Changed:**

- update trie-mux.
- add Log.KV method.

## [1.24.1] - 2021-06-11

**Changed:**

- add SetGraceTimeout option.
- clear mod.

## [1.24.0] - 2021-02-03

**Changed:**

- Remove ctx.Get and ctx.Set methods.
- Change Log structure fields: "time" -> "duration"; "timestamp" -> "time".

## [1.23.0] - 2021-01-07

**Changed:**

- Add app.ServeWithContext method.

## [1.22.0] - 2020-11-12

**Changed:**

- Update dependence.

## [1.21.7] - 2020-09-25

**Changed:**

- Add gear.MIMEApplicationSchemaGraphQL constant.
- Enlarge Error' Stack trace buffer size.

**Fixed:**

- Fix ctx.ParseBody' error code.

## [1.21.6] - 2020-08-28

**Changed:**

- Add gear.ErrorResponse struct that represents error response like JSON-RPC2 or Google cloud API.
- Add gear.RenderErrorResponse function that render error with ErrorResponse struct.

**Fixed:**

- Fix default SetRenderError function.
- Fix `xCanary` field in logging middleware.

## [1.21.5] - 2020-08-14

**Changed:**

- Add gear.SetRenderError hook that convert error to raw response.
- Add gear.MIMEApplicationSchemaJSONLD MIME type.

## [1.21.4] - 2020-08-05

**Changed:**

- Add default ReadHeaderTimeout, ReadTimeout, WriteTimeout and IdleTimeout for Server.

## [1.21.3] - 2020-07-31

**Fixed:**

- Fix ctx.ParseURL for empty query.

## [1.21.2] - 2020-05-14

**Fixed:**

- Fix ctx.WithContext: sets the context to underlying request object.

## [1.21.1] - 2020-03-28

**Changed:**

- Add ctx.Scheme() instead of ctx.Protocol().

**Fixed:**

- Fix ctx.Protocol() and ctx.IP() for proxy.

## [1.21.0] - 2020-03-24

**Changed:**

- Add request body and response body to access log when 500 error.

**Fixed:**

- Fix start time error for development logging.

## [1.20.0] - 2020-03-24

**Changed:**

- Add "X-Canary" Header.
- Add ctx.StartAt property.
- Add ctx.GetHeaders method.

## [1.19.0] - 2020-02-29

**Changed:**

- Add gear.App.ListenWithContext method.
- Add gear.ContextWithSignal function.
- Add logging.SetLoggerLevel function.

## [1.18.0] - 2020-02-23

**Changed:**

- Improve misdirected response with message (421).

**Fixed:**

- Fix error's stack.

## [1.17.1] - 2019-12-23

**Fixed:**

- Fix error's stack.

## [1.17.0] - 2019-09-03

**Changed:**

- Remove "application/x-www-form-urlencoded" supporting from DefaultBodyParser.

## [1.16.2] - 2019-07-29

**Fixed:**

- Fix logging middleware: fix time format.
- Fix logging middleware: change severity level to lowercase (https://en.wikipedia.org/wiki/Syslog).

## [1.16.1] - 2019-07-25

**Fixed:**

- Fix logging middleware: support JSON log.

## [1.16.0] - 2019-07-24

**Changed:**

- Update logging middleware: support JSON log.

## [1.15.4] - 2019-07-03

**Fixed:**

- Fix DefaultBodyParser's mediaType.

## [1.15.3] - 2019-06-13

**Fixed:**

- Allow add "after hook" and "end hook" after context ended but before header wrote.

## [1.15.2] - 2019-02-07

**Fixed:**

- Panic in "end hook" will be recovered.

## [1.15.1] - 2019-02-07

**Fixed:**


util.go
package gear

import (
	"bytes"
	"compress/gzip"
	"compress/zlib"
	"context"
	"encoding"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/textproto"
	"net/url"
	"os"
	"os/signal"
	"reflect"
	"runtime"
	"strconv"
	"strings"
	"sync/atomic"
	"syscall"
	"unicode/utf8"
)

type middlewares []Middleware

func (m middlewares) run(ctx *Context) (err error) {
	for _, fn := range m {
		if err = fn(ctx); !IsNil(err) || ctx.Res.ended.isTrue() {
			return
		}
	}
	return
}

// Compose composes a slice of middlewares to one middleware
func Compose(mds ...Middleware) Middleware {
	switch len(mds) {
	case 0:
		return noOp
	case 1:
		return mds[0]
	default:
		return middlewares(mds).run
	}
}

var noOp Middleware = func(ctx *Context) error { return nil }

// WrapHandler wrap a http.Handler to Gear Middleware
func WrapHandler(handler http.Handler) Middleware {
	return func(ctx *Context) error {
		handler.ServeHTTP(ctx.Res, ctx.Req)
		return nil
	}
}

// WrapHandlerFunc wrap a http.HandlerFunc to Gear Middleware
func WrapHandlerFunc(fn http.HandlerFunc) Middleware {
	return func(ctx *Context) error {
		fn(ctx.Res, ctx.Req)
		return nil
	}
}

// IsNil checks if a specified object is nil or not, without failing.
func IsNil(val any) bool {
	if val == nil {
		return true
	}

	value := reflect.ValueOf(val)
	switch value.Kind() {
	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.Interface, reflect.Slice:
		return value.IsNil()
	default:
		return false
	}
}

// Error represents a numeric error with optional meta. It can be used in middleware as a return result.
type Error struct {
	Code  int    `json:"-"`
	Err   string `json:"error"`
	Msg   string `json:"message"`
	Data  any    `json:"data,omitempty"`
	Stack string `json:"-"`
}

// ErrorResponse represents error response like JSON-RPC2 or Google cloud API.
type ErrorResponse struct {
	Error struct {
		Code    int    `json:"code"`
		Status  string `json:"status"`
		Message string `json:"message"`
		Data    any    `json:"data,omitempty"`
	} `json:"error"`
}

// ToErrorResponse convert a error to ErrorResponse instance.
func ToErrorResponse(e error) ErrorResponse {
	res := ErrorResponse{}
	err := Err.From(e)
	res.Error.Code = err.Code
	res.Error.Status = err.Err
	res.Error.Message = err.Msg
	res.Error.Data = err.Data
	return res
}

// errorForLog use to marshal for logging.
type errorForLog struct {
	Code  int    `json:"code"`
	Err   string `json:"error"`
	Msg   string `json:"message"`
	Data  any    `json:"data,omitempty"`
	Stack string `json:"stack"`
}

// Status implemented HTTPError interface.
func (err *Error) Status() int {
	return err.Code
}

// Error implemented HTTPError interface.
func (err *Error) Error() string {
	return fmt.Sprintf("%s: %s", err.Err, err.Msg)
}

// String implemented fmt.Stringer interface.
func (err Error) String() string {
	return err.GoString()
}

// GoString implemented fmt.GoStringer interface, returns a Go-syntax string.
func (err Error) GoString() string {
	if v, ok := err.Data.([]byte); ok && utf8.Valid(v) {
		err.Data = string(v)
	}
	return fmt.Sprintf(`Error{Code:%d, Err:%s, Msg:%s, Data:%#v, Stack:%s}`,
		err.Code, strconv.Quote(err.Err), strconv.Quote(err.Msg), err.Data, strconv.Quote(err.Stack))
}

// Format implemented logging.Messager interface.
func (err Error) Format() (string, error) {
	errlog := errorForLog(err)
	res, e := json.Marshal(errlog)
	if e == nil {
		return string(res), nil
	}
	return "", e
}

// WithErr returns a copy of err with given new error name.
//
//	err := gear.ErrBadRequest.WithErr("InvalidEmail") // 400 Bad Request error with error name InvalidEmail"
func (err Error) WithErr(name string) *Error {
	err.Err = name
	return &err
}

// WithMsg returns a copy of err with given new messages.
//
//	err := gear.Err.WithMsg() // just clone
//	err := gear.ErrBadRequest.WithMsg("invalid email") // 400 Bad Request error with message invalid email"
func (err Error) WithMsg(msgs ...string) *Error {
	if len(msgs) > 0 {
		err.Msg = strings.Join(msgs, ", ")
	}
	return &err
}

// WithMsgf returns a copy of err with given message in the manner of fmt.Printf.
//
//	err := gear.ErrBadRequest.WithMsgf(`invalid email: "%s"`, email)
func (err Error) WithMsgf(format string, args ...any) *Error {
	return err.WithMsg(fmt.Sprintf(format, args...))
}

// WithCode returns a copy of err with given code.
//
//	BadRequestErr := gear.Err.WithCode(400)
func (err Error) WithCode(code int) *Error {
	err.Code = code
	if text := http.StatusText(code); text != "" {
		err.Err = text
	}
	return &err
}

// WithStack returns a copy of err with error stack.
//
//	err := gear.Err.WithMsg("some error").WithStack()
func (err Error) WithStack(skip ...int) *Error {
	return ErrorWithStack(&err, skip...)
}

// From returns a copy of err with given error. It will try to merge the given error.
// If the given error is a *Error instance, it will be returned without copy.
//
//	err := gear.ErrBadRequest.From(errors.New("invalid email"))
//	err := gear.Err.From(someErr)
func (err Error) From(e error) *Error {
	if IsNil(e) {
		return nil
	}

	switch v := e.(type) {
	case *Error:
		return v
	case HTTPError:
		err.Code = v.Status()
		err.Msg = v.Error()
	case *textproto.Error:
		err.Code = v.Code
		err.Msg = v.Msg
	default:
		err.Msg = e.Error()
	}

	if err.Err == "" {
		err.Err = http.StatusText(err.Code)
	}
	return &err
}

// ParseError parse a error, textproto.Error or HTTPError to HTTPError
func ParseError(e error, code ...int) HTTPError {
	if IsNil(e) {
		return nil
	}

	switch v := e.(type) {
	case HTTPError:
		return v
	case *textproto.Error:
		err := Err.WithCode(v.Code)
		err.Msg = v.Msg
		return err
	default:
		err := ErrInternalServerError.WithMsg(e.Error())
		if len(code) > 0 && code[0] > 0 {
			err = err.WithCode(code[0])
		}
		return err
	}
}

// ErrorWithStack create a error with stacktrace
func ErrorWithStack(val any, skip ...int) *Error {
	if IsNil(val) {
		return nil
	}

	var err *Error
	switch v := val.(type) {
	case *Error:
		err = v.WithMsg() // must clone, should not change the origin *Error instance
	case error:
		err = ErrInternalServerError.From(v)
	case string:
		err = ErrInternalServerError.WithMsg(v)
	default:
		err = ErrInternalServerError.WithMsgf("%#v", v)
	}

	if err.Stack == "" {
		buf := make([]byte, 8192)
		buf = buf[:runtime.Stack(buf, false)]
		s := 0
		if len(skip) != 0 {
			s = skip[0]
		}
		err.Stack = pruneStack(buf, s)
	}
	return err
}

// ValuesToStruct converts url.Values into struct object. It supports specific types that implementing encoding.TextUnmarshaler interface.
//
//	type jsonQueryTemplate struct {
//		ID   string `json:"id" form:"id"`
//		Pass string `json:"pass" form:"pass"`
//	}
//
//	target := jsonQueryTemplate{}
//
//	gear.ValuesToStruct(map[string][]string{
//		"id": []string{"some id"},
//		"pass": []string{"some pass"},
//	}, &target, "form")
func ValuesToStruct(values map[string][]string, target any, tag string) (err error) {
	if values == nil {
		return fmt.Errorf("invalid values: %v", values)
	}
	if len(values) == 0 {
		return
	}
	rv := reflect.ValueOf(target)
	if rv.Kind() != reflect.Ptr || rv.IsNil() {
		return fmt.Errorf("invalid struct: %v", rv)
	}

	return valuesToStruct(values, rv, tag)
}

func valuesToStruct(values map[string][]string, rv reflect.Value, tag string) (err error) {
	rv = rv.Elem()
	rt := rv.Type()
	n := rv.NumField()
	for i := 0; i < n; i++ {
		structField := rt.Field(i)
		value := rv.Field(i)
		if structField.Anonymous {
			// embedded field
			if value.Kind() == reflect.Struct && value.CanAddr() {
				if err = valuesToStruct(values, value.Addr(), tag); err != nil {
					return
				}
			}
			continue
		}

		if !value.CanSet() {
			continue
		}

		fk := structField.Tag.Get(tag)
		if fk == "" {
			continue
		}

		if vals, ok := values[fk]; ok {
			if value.Kind() == reflect.Slice {
				err = setRefSlice(value, vals)
			} else if len(vals) > 0 && vals[0] != "" {
				err = setRefField(value, vals[0])
			}
			if err != nil {
				return
			}
		}
	}

	return
}

func shouldDeref(k reflect.Kind) bool {
	switch k {
	case reflect.String, reflect.Bool, reflect.Float32, reflect.Float64,
		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return true
	default:
		return false
	}
}

func setRefSlice(v reflect.Value, vals []string) error {
	if len(vals) == 1 && vals[0] != "" {
		if ok, err := tryUnmarshalValue(v, vals[0]); ok {
			return err
		}
	}

	l := len(vals)
	slice := reflect.MakeSlice(v.Type(), l, l)

	for i := 0; i < l; i++ {
		if err := setRefField(slice.Index(i), vals[i]); err != nil {
			return err
		}
	}

	v.Set(slice)
	return nil
}

func setRefField(v reflect.Value, str string) error {
	if ok, err := tryUnmarshalValue(v, str); ok {
		return err
	}

	if v.Kind() == reflect.Ptr && shouldDeref(v.Type().Elem().Kind()) {
		v.Set(reflect.New(v.Type().Elem()))
		v = v.Elem()
	}

	switch v.Kind() {
	case reflect.String:
		v.SetString(str)
		return nil
	case reflect.Bool:
		return setRefBool(v, str)
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return setRefInt(v, str, v.Type().Bits())
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return setRefUint(v, str, v.Type().Bits())
	case reflect.Float32, reflect.Float64:
		return setRefFloat(v, str, v.Type().Bits())
	default:
		return fmt.Errorf("unknown field type: %v", v.Type())
	}
}

func setRefBool(v reflect.Value, str string) error {
	val, err := strconv.ParseBool(str)
	if err == nil {
		v.SetBool(val)
	}
	return err
}

func setRefInt(v reflect.Value, str string, size int) error {
	val, err := strconv.ParseInt(str, 10, size)
	if err == nil {
		v.SetInt(val)
	}
	return err
}

func setRefUint(v reflect.Value, str string, size int) error {
	val, err := strconv.ParseUint(str, 10, size)
	if err == nil {
		v.SetUint(val)
	}
	return err
}

func setRefFloat(v reflect.Value, str string, size int) error {
	val, err := strconv.ParseFloat(str, size)
	if err == nil {
		v.SetFloat(val)
	}
	return err
}

func tryUnmarshalValue(v reflect.Value, str string) (bool, error) {
	if v.Kind() != reflect.Ptr && v.CanAddr() && v.Type().Name() != "" {
		v = v.Addr()
	}

	if v.Type().NumMethod() > 0 {
		if v.IsNil() {
			v.Set(reflect.New(v.Type().Elem()))
		}
		if u, ok := v.Interface().(encoding.TextUnmarshaler); ok {
			return true, u.UnmarshalText([]byte(str))
		}
	}
	return false, nil
}

// pruneStack make a thin conversion for stack information
// limit the count of lines to 5
// src:
// ```
// goroutine 9 [running]:
// runtime/debug.Stack(0x6, 0x6, 0xc42003c898)
//
//	/usr/local/Cellar/go/1.7.4_2/libexec/src/runtime/debug/stack.go:24 +0x79
//
// github.com/teambition/gear/logging.(*Logger).OutputWithStack(0xc420012a50, 0xed0092215, 0x573fdbb, 0x471f20, 0x0, 0xc42000dc1a, 0x6, 0xc42000dc01, 0xc42000dca0)
//
//	/Users/xus/go/src/github.com/teambition/gear/logging/logger.go:267 +0x4e
//
// github.com/teambition/gear/logging.(*Logger).Emerg(0xc420012a50, 0x2a9cc0, 0xc42000dca0)
//
//	/Users/xus/go/src/github.com/teambition/gear/logging/logger.go:171 +0xd3
//
// github.com/teambition/gear/logging.TestGearLogger.func2(0xc420018600)
//
//	/Users/xus/go/src/github.com/teambition/gear/logging/logger_test.go:90 +0x3c1
//
// testing.tRunner(0xc420018600, 0x33d240)
//
//	/usr/local/Cellar/go/1.7.4_2/libexec/src/testing/testing.go:610 +0x81
//
// created by testing.(*T).Run
//
//	/usr/local/Cellar/go/1.7.4_2/libexec/src/testing/testing.go:646 +0x2ec
//
// ```
// dst:
// ```
// Stack:
//
//	/usr/local/Cellar/go/1.7.4_2/libexec/src/runtime/debug/stack.go:24
//	/Users/xus/go/src/github.com/teambition/gear/logging/logger.go:283
//	/Users/xus/go/src/github.com/teambition/gear/logging/logger.go:171
//	/Users/xus/go/src/github.com/teambition/gear/logging/logger_test.go:90
//	/usr/local/Cellar/go/1.7.4_2/libexec/src/testing/testing.go:610
//	/usr/local/Cellar/go/1.7.4_2/libexec/src/testing/testing.go:646
//
// ```
func pruneStack(stack []byte, skip int) string {
	// remove first line
	// `goroutine 1 [running]:`
	lines := strings.Split(string(stack), "\n")[1:]
	newLines := make([]string, 0, len(lines)/2)

	num := 0
	for idx, line := range lines {
		if idx%2 == 0 {
			continue
		}
		skip--
		if skip >= 0 {
			continue
		}
		num++

		loc := strings.Split(line, " ")[0]
		loc = strings.Replace(loc, "\t", "\\t", -1)
		// only need odd line
		newLines = append(newLines, loc)
		if num == 10 {
			break
		}
	}
	return strings.Join(newLines, "\\n")
}

type atomicBool int32

func (b *atomicBool) isTrue() bool {
	return atomic.LoadInt32((*int32)(b)) == 1
}

func (b *atomicBool) swapTrue() bool {
	return atomic.SwapInt32((*int32)(b), 1) == 0
}

func (b *atomicBool) setTrue() {
	atomic.StoreInt32((*int32)(b), 1)
}

// IsStatusCode returns true if status is HTTP status code.
// https://en.wikipedia.org/wiki/List_of_HTTP_status_codes
// https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
func IsStatusCode(status int) bool {
	switch status {
	case 100, 101, 102, 103,
		200, 201, 202, 203, 204, 205, 206, 207, 208, 226,
		300, 301, 302, 303, 304, 305, 306, 307, 308,
		400, 401, 402, 403, 404, 405, 406, 407, 408, 409, 410, 411, 412, 413, 414, 415, 416, 417, 418,
		421, 422, 423, 424, 426, 428, 429, 431, 440, 444, 449, 450, 451, 494, 495, 496, 497, 498, 499,
		500, 501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511, 520, 521, 522, 523, 524, 525, 526, 527:
		return true
	default:
		return false
	}
}

func isRedirectStatus(status int) bool {
	switch status {
	case 300, 301, 302, 303, 305, 307, 308:
		return true
	default:
		return false
	}
}

func isEmptyStatus(status int) bool {
	switch status {
	case 100, 101, 102, 103, 204, 205, 304:
		return true
	default:
		return false
	}
}

// ContentDisposition implements a simple version of https://tools.ietf.org/html/rfc2183
// Use mime.ParseMediaType to parse Content-Disposition header.
func ContentDisposition(fileName, dispositionType string) (header string) {
	if dispositionType == "" {
		dispositionType = "attachment"
	}
	if fileName == "" {
		return dispositionType
	}

	header = fmt.Sprintf(`%s; filename="%s"`, dispositionType, url.QueryEscape(fileName))
	fallbackName := url.PathEscape(fileName)
	if len(fallbackName) != len(fileName) {
		header = fmt.Sprintf(`%s; filename*=UTF-8''%s`, header, fallbackName)
	}
	return
}

// LoggerFilterWriter is a writer for Logger to filter bytes.
// In a https server, avoid some handshake mismatch condition such as loadbalance healthcheck:
//
//	2017/06/09 07:18:04 http: TLS handshake error from 10.10.5.1:45001: tls: first record does not look like a TLS handshake
//	2017/06/14 02:39:29 http: TLS handshake error from 10.0.1.2:54975: read tcp 10.10.5.22:8081->10.0.1.2:54975: read: connection reset by peer
//
// Usage:
//
//	func main() {
//		app := gear.New() // Create app
//		app.Set(gear.SetLogger, log.New(gear.DefaultFilterWriter(), "", 0))
//		app.Use(func(ctx *gear.Context) error {
//			return ctx.HTML(200, "<h1>Hello, Gear!</h1>")
//		})
//
//		app.Listen(":3000")
//	}
type LoggerFilterWriter struct {
	phrases [][]byte
	out     io.Writer
}

var loggerFilterWriter = &LoggerFilterWriter{
	phrases: [][]byte{[]byte("http: TLS handshake error"), []byte("EOF")},
	out:     os.Stderr,
}

// DefaultFilterWriter returns the default LoggerFilterWriter instance.
func DefaultFilterWriter() *LoggerFilterWriter {
	return loggerFilterWriter
}

// SetOutput sets the output destination for the loggerFilterWriter.
func (s *LoggerFilterWriter) SetOutput(out io.Writer) {
	s.out = out
}

// Add add a phrase string to filter
func (s *LoggerFilterWriter) Add(err string) {
	if s.out == nil {
		panic(Err.WithMsg("output io.Writer should be set with SetOutput method"))
	}
	s.phrases = append(s.phrases, []byte(err))
}

func (s *LoggerFilterWriter) Write(p []byte) (n int, err error) {
	for _, phrase := range s.phrases {
		if bytes.Contains(p, phrase) {
			return len(p), nil
		}
	}

	return s.out.Write(p)
}

// Decompress wrap the reader for decompressing, It support gzip and zlib, and compatible for deflate.
func Decompress(encoding string, r io.Reader) (io.ReadCloser, error) {
	switch encoding {
	case "gzip":
		return gzip.NewReader(r)
	case "deflate", "zlib":
		// compatible for RFC 1950 zlib, RFC 1951 deflate, http://www.open-open.com/lib/view/open1460866410410.html
		return zlib.NewReader(r)
	default:
		return nil, ErrUnsupportedMediaType.WithMsgf("Unsupported Content-Encoding: %s", encoding)
	}
}

// https://tools.ietf.org/html/rfc6838
// https://www.iana.org/assignments/media-types/media-types.xml
// application/jrd+json, application/jose+json, application/geo+json, application/geo+json-seq and so on.
func isLikeMediaType(s, t string) bool {
	if !strings.HasPrefix(s, "application/") {
		return false
	}
	switch t {
	case "json":
		return strings.Contains(s, "+json")
	case "xml":
		return strings.Contains(s, "+xml")
	default:
		return false
	}
}

// ContextWithSignal creates a context canceled when SIGINT or SIGTERM are notified
//
// Usage:
//
//	 func main() {
//	 	app := gear.New() // Create app
//	 	do some thing...
//
//	 	app.ListenWithContext(gear.ContextWithSignal(context.Background()), addr)
//		  // starts the HTTPS server.
//		  // app.ListenWithContext(gear.ContextWithSignal(context.Background()), addr, certFile, keyFile)
//	 }
func ContextWithSignal(ctx context.Context) context.Context {
	newCtx, cancel := context.WithCancel(ctx)
	signals := make(chan os.Signal)
	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
	go func() {
		<-signals
		cancel()
	}()
	return newCtx
}

// RenderErrorResponse is a SetRenderError function with ErrorResponse struct.
// It will become a default SetRenderError function in Gear@v2.
//
// Usage:
//
//	app.Set(gear.SetRenderError, gear.RenderErrorResponse)
func RenderErrorResponse(err HTTPError) (int, string, []byte) {
	body, e := json.Marshal(ToErrorResponse(err))
	if e != nil {
		body, _ = json.Marshal(map[string]any{
			"error": map[string]string{"message": err.Error()},
		})
	}
	return err.Status(), MIMEApplicationJSONCharsetUTF8, body
}

func defaultRenderError(err HTTPError) (int, string, []byte) {
	// default to render error as json
	body, e := json.Marshal(err)
	if e != nil {
		body, _ = json.Marshal(map[string]string{"error": err.Error()})
	}
	return err.Status(), MIMEApplicationJSONCharsetUTF8, body
}
